pandas/tests/reshape/merge/test_merge.py

```diff
@@ -1369,12 +1364,118 @@ def f():
+
+
+@pytest.fixture
+def on_cols():
+    return ['Origin', 'Destination', 'Period']
 
-        pytest.raises(NotImplementedError, f)
+
+@pytest.fixture
+def idx_cols():
+    return ['Origin', 'Destination', 'Period', 'TripPurp', 'LinkType']
+
+
+class TestJoinMultiMulti(object):
+
+    @pytest.mark.parametrize('how', ['left', 'right', 'inner', 'outer'])
+    def test_join_multi_multi(self, left_multi, right_multi, how,
+                              on_cols, idx_cols):
+        # Multi-index join tests
+        expected = (pd.merge(left_multi.reset_index(),
+                             right_multi.reset_index(),
+                             how=how, on=on_cols).set_index(idx_cols)
+                    .sort_index())
+
+        result = left_multi.join(right_multi, how=how).sort_index()
+        tm.assert_frame_equal(result, expected)
+
+    """
+    @pytest.mark.parametrize('how', ['left', 'right', 'inner', 'outer'])
+    def test_join_multi_multi_emptylevel(self, left_multi, right_multi, how,
+                                         on_cols, idx_cols):
+        # Join with empty level
+        num_lvls = len(right_multi.index.get_level_values('Period'))
+        # Set one level to None
+        right_multi.index.set_levels([np.nan] * num_lvls, level='Period',
+                                     inplace=True)
+
+        expected = (pd.merge(left_multi.reset_index(),
+                             right_multi.reset_index(),
+                             how=how, on=on_cols).set_index(idx_cols)
+                    .sort_index())
+
+        result = left_multi.join(right_multi, how=how).sort_index()
+        tm.assert_frame_equal(result, expected)
+    """
+
+    @pytest.mark.parametrize('how', ['left', 'right', 'inner', 'outer'])
+    def test_join_multi_empty_frames(self, left_multi, right_multi, how,
+                                     on_cols, idx_cols):
+
+        left_multi = left_multi.drop(columns=left_multi.columns)
+        right_multi = right_multi.drop(columns=right_multi.columns)
+
+        expected = (pd.merge(left_multi.reset_index(),
+                             right_multi.reset_index(),
+                             how=how, on=on_cols).set_index(idx_cols)
+                    .sort_index())
+
+        result = left_multi.join(right_multi, how=how).sort_index()
+        tm.assert_frame_equal(result, expected)
+
+    def test_join_multi_multi_nonunique(self, left_multi):
+        # Non-unique resulting index
+        right_multi = (
+            DataFrame(
+                dict(Origin=[1, 1, 2],
+                     Destination=[1, 1, 1],
+                     Period=['AM', 'AM', 'PM'],
+                     LinkType=['a', 'b', 'a'],
+                     Distance=[100, 110, 120]),
+                columns=['Origin', 'Destination', 'Period',
+                         'LinkType', 'Distance'])
+            .set_index(['Origin', 'Destination', 'Period', 'LinkType']))
+
+        def f():
```

Comment:
are there other error conditions to test?

Reply:
This test fails and I'm not sure how we should handle join on empty levels

https://github.com/harisbal/pandas/blob/d694f254f34d3a8bb1ccdbafbc876d5f28ea3323/pandas/tests/reshape/merge/test_merge.py#L1429-L1446

Reply:
This test fails and I'm not sure how we should handle join on empty levels

https://github.com/harisbal/pandas/blob/d694f254f34d3a8bb1ccdbafbc876d5f28ea3323/pandas/tests/reshape/merge/test_merge.py#L1429-L1446

Reply:
can you show a mini-example?

Reply:
@jreback Here is an example of a join on 2 multilevel indexed (same levels) dfs using two different methods
1. pd.merge(df1.reset_index(), df2.reset_index(),...)
2. df1.join(df2)

The results differ. Do you think that's an issue? I'm facing a similar issue when I try a multi-level join.

import numpy as np
import pandas as pd
import pandas.util.testing as tm

join_type='left'

dests=[np.nan, np.nan, np.nan, np.nan, np.nan]
left_multi=(
        pd.DataFrame(
            dict(Origin=['A', 'A', 'B', 'B', 'C'],
                 Destination=dests,
                 Trips=[1987, 3647, 2470, 4296, 4444]),
            columns=['Origin', 'Destination', 'Trips'])
        .set_index(['Origin', 'Destination']))

dests=[np.nan, np.nan, np.nan, np.nan, np.nan, np.nan, np.nan]
right_multi=(
        pd.DataFrame(
            dict(Origin=['A', 'A', 'B', 'B', 'C', 'C', 'E'],
                 Destination=dests,
                 Distance=[100, 80, 90, 80, 75, 35, 55]),
            columns=['Origin', 'Destination', 'Distance'])
        .set_index(['Origin', 'Destination']))


on_cols = ['Origin', 'Destination']
idx_cols = ['Origin', 'Destination']

expected = (pd.merge(left_multi.reset_index(),
                     right_multi.reset_index(),
                     how=join_type, on=on_cols).set_index(idx_cols)
            .sort_index())

result = left_multi.join(right_multi, how=join_type).sort_index()

print(expected)
print(result)

tm.assert_frame_equal(result, expected)